pandas/_libs/lib.pyx

```diff
@@ -2359,3 +2356,106 @@ def fast_multiget(dict mapping, ndarray keys, default=np.nan):
             output[i] = default
 
     return maybe_convert_objects(output)
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+cdef inline void convert_and_set_item(object item, Py_ssize_t index,
+                                      object[:] result,
+                                      bint keep_trivial_numbers):
+    cdef:
+        bint do_convert = 1
+        double double_item
+
+    if keep_trivial_numbers:
+        if isinstance(item, int) and Py_SIZE(item) < 2:
+            if <int>item == 0:
+                do_convert = 0
+        elif isinstance(item, float):
+            double_item = item
+            if double_item == 0.0 or double_item != double_item:
+                do_convert = 0
+
+    if do_convert and not isinstance(item, (str, bytes)):
+        item = PyObject_Str(item)
+
+    result[index] = item
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+cdef inline void put_object_as_unicode(list lst, Py_ssize_t idx,
```

Comment:
why is this a separate function?

Reply:
To not duplicate same code for numpy arrays and other iterables.
We're now making an experiment to see if there's big difference, and if there isn't one we'll inline this function.